tests/unit/locking_test.cc

```diff
@@ -420,3 +422,170 @@ SEASTAR_THREAD_TEST_CASE(test_with_lock_typed_return_throwing_move_func) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    try {
+        const auto ulock = co_await get_unique_lock(sm);
+        throw std::runtime_error("injected");
+    } catch (const std::runtime_error&) {
+        BOOST_REQUIRE(sm.try_lock());
+        sm.unlock();
+    } catch (...) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_lock_locks) {
+#ifdef SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+    shared_mutex sm;
+
+    // If l.lock() fails, neither the function, nor l.unlock()
+    // should be called.
+    const auto ulock = co_await get_unique_lock(sm);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await std::invoke([&] () -> future<> {
+        const auto slock = co_await get_shared_lock(sm);
+        BOOST_REQUIRE(false);
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await std::invoke([&] () -> future<> {
+        const auto ulock = co_await get_unique_lock(sm);
+        BOOST_REQUIRE(false);
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().cancel();
+#endif // SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+
+    // If the macro resolves to 0, we need this to prevent compilation errors
+    // since we're not returning anything from this function.
+    co_return;
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_lock_mixed_locks) {
+#ifdef SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+    shared_mutex sm;
+
+    // if l.lock() fails neither the function nor l.unlock()
+    // should be called.
+    const auto ulock = co_await get_unique_lock(sm);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await with_shared(sm, [] () -> future<> {
+        BOOST_REQUIRE(false);
+        co_return;
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await with_lock(sm, [] () -> future<> {
+        BOOST_REQUIRE(false);
+        co_return;
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().cancel();
+#endif // SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+
+    // If the macro resolves to 0, we need this to prevent compilation errors
+    // since we're not returning anything from this function.
+    co_return;
+}
+
+SEASTAR_TEST_CASE(test_with_shared_typed_return_nothrow_move_func_locks) {
+    shared_mutex sm;
+
+    int expected = 42;
+    auto res = co_await std::invoke([&sm, &expected] () -> future<int> {
+        const auto slock = co_await get_shared_lock(sm);
+        co_return expected;
+    });
+    BOOST_REQUIRE_EQUAL(res, expected);
+
+    try {
+        const auto slock = co_await get_shared_lock(sm);
+        if (expected == 42) {
+            throw expected_exception(expected);
```

Comment:
In test_shared_mutex_failed_func_locks above you already tested that an exception correctly unlocks the mutex. This seems to test the same thing, but not as well (you didn't check that sm got unlocked!). So why do we need this test? Why does this test's name have the strings "return_nothrow_move" - what does this refer to?

Reply:
Good catch. I forgot the rename the test. It was originally doing something else (and as most of the other tests I added, it was based on the existing ones). But you're right that it's unnecessary. I'll remove it.